### DISCUSSION TITLE
Support for native RS485 for TCP-Serial-Bridge

### DISCUSSION THREAD

**Discussion by TomsTek** [arendst/Tasmota#21238] [Ideas]
## Overview

**Add support for using MAX485 or equivalent to use TCP serial bridge with RS485 networks.**

![image](https://github.com/arendst/Tasmota/assets/51259714/e814531e-ce71-4b58-a1c4-89b8e1d78d32)

(It is an equivalent function as Modbus Bridge. (xdrv_63_modbus_bridge.ino, TasmotaModbus.cpp))

I tested the functionality of my ESP32 board with the following changes.

Tasmota version: 13.4.0 (tasmota32) (self-compiled version)
 
However, I'm not sure if I can implement the feature correctly in the project, so I need help.

## First step: Add pin definition GPIO_TCP_TX_EN

Changes in be_gpio_defines.h
![image](https://github.com/arendst/Tasmota/assets/51259714/b5e563be-5a4d-4c54-8ac8-3e763fac958a)
Added lines
“
{ "TCP_TX_EN", (int32_t) GPIO_TCP_TX_EN },
“

I'm not sure if it's necessary to add a line in tasmota_template_legacy.h

Changes in tasmota_template.h
![image](https://github.com/arendst/Tasmota/assets/51259714/0d3b3eea-4ab7-4343-9f95-e490ec01c751)
Added lines
```
  GPIO_TCP_TX_EN,                       // TCP to serial bridge
```
![image](https://github.com/arendst/Tasmota/assets/51259714/79b1479e-aacd-4392-ac48-53315abaaf12)
Added lines
```
  AGPIO(GPIO_TCP_TX_EN),                   // TCP Serial bridge
```

Changes in lv_gpio_enum.h
![image](https://github.com/arendst/Tasmota/assets/51259714/576005f3-2752-4720-9e3c-0d1e2afee116)
Added lines
```
TCP_TX_EN = GPIO_TCP_TX_EN
```

The pin selector worked but has no text. So there must be an error there.
![image](https://github.com/arendst/Tasmota/assets/51259714/9cd1e942-48ab-4fdb-a3f4-307c16c27601)

## second Step: add RS485 DE-Pin functionality

Changes in xdrv_41_tcp_bridge.ino
![image](https://github.com/arendst/Tasmota/assets/51259714/a03b4236-5224-46ac-ae7f-8d80ef9f2230)
Added lines
```
    if (PinUsed(GPIO_TCP_TX_EN)) {
      TCPSerial->setTransmitEnablePin(Pin(GPIO_TCP_TX_EN));
      AddLog(LOG_LEVEL_DEBUG, PSTR(D_LOG_TCP "TCP Bridge DE is used on Pin: %d"), Pin(GPIO_TCP_TX_EN));
    }
```
The log line is optional

Thank you in advance for support in implementing the functionality.

**Comment by s-hadinger** [comment]
Leave it to me. Thanks

**Comment by TomsTek** [comment]
Is it an idea to make RS485 (controll Driver Enable Pin) available with Berry?

**Reply by barbudor** [reply]
The C++ class `TasmotaSerial` on which Berry's `serial` class is built upon has a member `setTransmitEnablePin` to define the GPIO used for that. It is for exemple used in the Tasmota ModBus driver.

It should be fairly simple to extend the Berry's `serial` class to either
- add a `serial.setTransmitEnablePin()` member mapped to the C++ equivalent
- add an optional argument to the `serial()` contructor for that enable pin and do the call on object creation

What do you think @s-hadinger ?

**Reply by s-hadinger** [reply]
Sure, I prefer an additional method to serial class, like in C++

**Comment by s-hadinger** [comment]
It's merged, see https://github.com/arendst/Tasmota/pull/21269

**Reply by TomsTek** [reply]
Thanks! 
I try to test it this week.

**Comment by TomsTek** [comment]
Last weekend I tested the TCP Serial Bridge with ESP32 and ESP8266 and it worked well.

Thanks for the implementation.

**Comment by jpmorrison** [comment]
Good feature to have. How does it compare to newer 485 modules that do this in hardware? It saves a GPIO and works with a PZEM energy meter.  MAX13487 has a built in state machine/direction control - I haven't tested, could be fake or mislabelled MAX3485 (3.3V).

**Reply by TomsTek** [reply]
The MAX13487 really exists. Here is the data sheet from Maxim Integrated (now part of Analog Devices):
https://www.analog.com/media/en/technical-documentation/data-sheets/MAX13487E-MAX13488E.pdf

Most RS485 driver circuits that only use the two UART lines (Rx, Tx) have some kind of limitation. Only the microcontroller itself really knows whether it wants to send or receive.
Some circuits work with logic gates and timing elements to generate the information for the DE pin. The time depends on the baud rate, which you then need to know approximately. If that fits, it works too.

The MAX13487 works internally with a short time element. For this reason it is necessary that the bus lines have correct pullups and pulldowns. This is the only way to ensure that the bus has valid levels when it is not transmitting.
In addition, it always requires a 5V power supply and is an expensive RS485 driver.
![Image](https://github.com/arendst/Tasmota/assets/51259714/73ea4860-5e78-419f-814e-d484c696a9d3)

Regardless, this is a fascinating piece that I didn't know about before. ***Thanks for the information.*** And I think it's good to have the choice.

**Reply by TomsTek** [reply]
The THVD1406/THVD1426 solves the supply problem (3.3 to 5 V) and is cheaper than MAX13487.
https://www.ti.com/product/de-de/THVD1406

**Reply by jpmorrison** [reply]
I saw the datasheet - I was just skeptical about the MAX13487 pcb modules on Aliexpress. 
If they are real, the price difference for a finished board is negligible and includes TVS diodes.
If they are fake, maybe you lose $1?
